### ISSUE TITLE
Can't jump to bookmark location when the key is a digit

### ISSUE THREAD

**Issue by dangerware** [jarun/nnn#420]
This is the content of my `NNN_BMS` variable:
```
export NNN_BMS="w:/media/disk2/work;2:/media/usb2"
```
When running in interactive mode I can't jump to the directory that is assigned to the key `2` using `<Leader>2`. Instead it will select the second context. Yet if I execute it like `nnn -b 2` it starts `nnn` in the path defined by key `2`.

To get the expect behavior, I cloned this repo at [master](https://github.com/jarun/nnn/tree/f480de0065acdccb5afafa717665fe62b3110f43) and added the following after line [4646](https://github.com/jarun/nnn/blob/f480de0065acdccb5afafa717665fe62b3110f43/src/nnn.c#L4646) in `nnn.c`:
```c
if (sel == SEL_LEADER)
    break;
```

Should I turn this into a pull request?
Thank you!

Operating System: Arch Linux (LTS Kernel v4.19.91)
Desktop Environment: i3
Terminal Emulator: kitty
Shell: bash or fish
Issue exists on `nnn` master: yes (Initially found on v2.8.1)

**Comment by 0xACE** [comment]
https://github.com/jarun/nnn/blob/f480de0065acdccb5afafa717665fe62b3110f43/src/nnn.c#L4611-L4618
⋮
https://github.com/jarun/nnn/blob/f480de0065acdccb5afafa717665fe62b3110f43/src/nnn.c#L4643-L4647
⋮
https://github.com/jarun/nnn/blob/f480de0065acdccb5afafa717665fe62b3110f43/src/nnn.c#L4670-L4671

TL;DR: keys: `1-4` and a couple of others get special treatment before they get parsed by the bookmark handler. The `-b` flag does not: https://github.com/jarun/nnn/blob/f480de0065acdccb5afafa717665fe62b3110f43/src/nnn.c#L5780-L5782

I'm not familiar with this code. I'm not the right person to make the final call on this, as I don't know what the expected behavior is.

Someone who knows better will let you know soon.

**Comment by jarun** [comment]
I am thinking of removing the special powers of the lead key. This was initially added to help people in `nav-as-you-type` mode but one can also press `Esc` instead of the Lead key to do the same thing.

@0xACE what do you think?

**Comment by 0xACE** [comment]
> I am thinking of removing the special powers of the lead key. This was initially added to help people in `nav-as-you-type` mode but one can also press `Esc` instead of the Lead key to do the same thing.

This current design leads to inconsistent behavior. Besides `.` none of these additions actually provide a different function compared to `Esc`.

I'm not the right person to decide if they should be removed or not as I never used them from within `filterentries()`. Instead I added my own special keys within `filterentries()` to access those features. You should see with other users who may be affected by this.


I know i said I would publish my changes, but atm they are a bit of a cluster f*ck so I never cleaned up... but within filter entries I have these changes applied:
```
case 'K': case 'J': // go up/down in the filtered entries list without leaving nav-as-you-type
	if (ndents) { if (*ch == 'J') move_cursor((cur + 1) % ndents, 0); \
		else if (*ch == 'K') move_cursor((cur + ndents - 1) % ndents, 0); } \
	redraw(path); printprompt(ln); continue; \
case 'Q': *ch = 'q'; break; // I have mapped q to ESC in nnn, Q will let you type the letter q
case 'N': case '%': case '"': case 'G': case 'P': case 'E': case '!': case '`': case '@': case '~': case '#': case 'L': case 'H': case 'C': case 'q':
	if (*ch == 'q') *ch = 27; // q lets you quit prompts.
	else if (*ch == 'H') *ch = 'h'; // go up 1 dir
	else if (*ch == 'L') *ch = 'l'; // go into dir (or open file)
	else if (*ch == 'P') *ch = 'p'; // preview file
	else if (*ch == 'E') *ch = 'e'; // edit file rather than open
	else if (*ch == '#') *ch = '-'; // jump to last dir
// fallthrough to case 27: /* exit filter mode on escape */
```
*(Note: My filter entries is case-insensitive and not regex based. basically regular text search but each word delimited by `space` must be present, i find it helpful in 90% of my use cases)*

**Comment by jarun** [comment]
Let's leave the context handling as it is. I can see an additional advantage - you can retain filters if you switch the contexts using the Lead key.

**Comment by jarun** [comment]
I'm closing the issue as this is documented from a long time: https://github.com/jarun/nnn/wiki/Usage#lead-key

and the behaviour is as expected.

**Comment by jarun** [comment]
@0xACE why not document your changes in the wiki (like the pager one)? It will be useful to many users.

**Comment by 0xACE** [comment]
> @0xACE why not document your changes in the wiki (like the pager one)? It will be useful to many users.

are you referring to opening files in the pager from within `filterentries()` using `case 'P':` ?

It only works well because my `and_str()` is case-insensitive. i'm not sure if `visible_re()` is case-insentivie. judging by `visible_str()` I'm guessing both of these functions are case-sensitive

you could apply this patch locally if you want to try it out:

```
diff --git a/src/nnn.c b/src/nnn.c
index a420777..7baf2b8 100644
--- a/src/nnn.c
+++ b/src/nnn.c
@@ -2053,6 +2053,8 @@ static int filterentries(char *path)
 
 			printprompt(ln);
 			continue;
+		case 'P':
+			*ch = 'p'; // fallthrough
 		case KEY_MOUSE: // fallthrough
 		case 27: /* Exit filter mode on Escape */
 			if (len == 1)
```

**Comment by jarun** [comment]
I've change the Lead key to a bookmark key. Made the necessary modification to retain the filter on context change when in nav-as-you-type mode. 

This actually simplifies a lot of things.

**Comment by dangerware** [comment]
Thank you very much! It's working fine with that fix.

**Comment by jarun** [comment]
It's not a fix. I changed a well-defined behaviour. And I see issues with clearing filters. So I may revert it back.

Also, I don't see it as a significant issue that 1-4 are reserved. There's a-d and so on.

**Comment by jarun** [comment]
I had to revert this. I spent half a day but couldn't straighten the filter workflow without messing up the code.

**Comment by dangerware** [comment]
No problem. Thank you anyway for the time you've spent. It would be convenient to me but in the end it's no big deal.

**Comment by jarun** [comment]
To clarify the issue for others, we retain filters while changing contexts. If we rely on `Esc` to exit filter mode but retain the filter, the filter will show in later refreshes of when you reload a new directory (like jumping to a bookmark) if you use nav-as-you-type mode. To handle this we can have a check before dir reload to clear the filter, but that will be too many checks for the type of selection.

In the current scenario, the lead key can handle context switches so you are still in nav as you type mode and the filter is meaningful. Also we retain the filter only on Tab which continues in nav-as-you-type mode.

**Comment by jarun** [comment]
@dangerware I am hoping this is fixed for good now. I have simplified the workflows greatly. The bookmark key will now handle only bookmarks. The lead key is gone. `Esc` from _nav-as-you-type_ to jump to special keys. It took a few days to stabilize this. Please spend some time to test.

@0xACE @KlzXS @annagrram please use master for your regular work so it gets tested.

Please check the help settings for keybind changes. Lead key is gone so could make some sane changes now.

**Comment by dangerware** [comment]
@jarun First of all, thank you **very much** for getting back to this. I had already assumed the issue was settled. Tomorrow I'm going to pull the master branch, test it and let you know.

**Comment by dangerware** [comment]
As far as I can tell, this is working just fine! I've changed my `NNN_BMS` for both Bash and Fish to the following:

```
export NNN_BMS="1:/media/usb1;2:/media/usb2;3:/media/usb3;a:$ART;d:$DLS;j:$JUNK;p:$PIC;r://;u:$AUR;w:$WORK"
set --export NNN_BMS "1:/media/usb1;2:/media/usb2;3:/media/usb3;a:$ART;d:$DLS;j:$JUNK;p:$PIC;r://;u:$AUR;w:$WORK"
```

So all 10 slots are being used. Works fine under Arch Linux (LTS 4.19.91) + i3 + kitty.
Tested with the master branch at db647e7c4c0e482f9b7fb0e351abd4730d5b5cac.

Once again, thank you!

**Comment by jarun** [comment]
Thanks for confirming.

**Comment by jarun** [comment]
@psevdaisthisi symlinked bookmarks is now implemented in the core program. Will be available on the next release.
